### DISCUSSION TITLE
[lxqt-panel] Some ideas for the panel

### DISCUSSION THREAD

**Discussion by elviosak** [lxqt/lxqt#1991] [Ideas]
There are a few changes i want to make, so if you like it i plan to write them.

- Allow more than one `StatusNotifierWidget`

this would allow have statusnotifier on multiple panels, or on the same one for whatever reason.
not sure how KDE does it, but i already wrote a concept using a singleton, it's working well: https://github.com/slidinghotdog/lxqt-panel/tree/statusnotifier

- Allow all panel positions with some warnings when unsupported positions.

i made a patch to allow this, and have been using it for a while. i tested a few WMs and most of them have some weirdness when panel is between different screens, but with `xfwm4` i did not notice any problems, i remember some issues on `lxqt-panel`'s github when the option was removed, so if added ppl can use it if it's not causing problems. https://github.com/slidinghotdog/lxqt-panel/tree/position-override

here's how the config looks (main screen bottom, secondary screen on top):
![panel-override](https://user-images.githubusercontent.com/33790211/117524924-3db29100-af96-11eb-9294-497e8fc3da51.png)

- Add another `xfce4-whisker-menu` style menu, or change current one.

i really like the 2-pane menus, probably there's more ppl that want it?

- Tiling options when dragging windows

just an idea i had, when dragging buttons from taskbar, show some tiling options that get triggered if the window is dropped in it (kinda hard to explain, but when you start dragging a window button, show a window on the same position as the `LXQtTaskButton` menus with options for tiling left, right, left-top and so on)

**Comment by tsujan** [comment]
> Allow more than one StatusNotifierWidget

I think that's impossible.

> Allow all panel positions with some warnings when unsupported positions.

Would cause serious troubles that we had a hard time to avoid.

> Tiling options when dragging windows
 
Didn't understand this one.

**Reply by elviosak** [reply]
>> Tiling options when dragging windows

>Didn't understand this one.

ok i made an example, this is what i meant, it would have a grid with various configurable tiling options.


https://user-images.githubusercontent.com/33790211/124663178-243ba180-de80-11eb-8140-8e96e0aada44.mp4

**Comment by elviosak** [comment]
>> Allow more than one StatusNotifierWidget

>I think that's impossible.

in my example it works fine, i created a singleton proxy class to handle `StatusNotifierWatcher` and it forwards the signals to `StatusNotifierWidget`

>>Allow all panel positions with some warnings when unsupported positions.

>Would cause serious troubles that we had a hard time to avoid.

my idea was to have the option but make it clear it "can" cause problems, but since i don't have problems with `xfwm4` i can still place the panels anywhere
in the patch i just added (Unsupported) to the options, as u can see on the screenshot, but i could add more information or add some steps like a checkbox to enable unsupported positions.

>>Tiling options when dragging windows

>Didn't understand this one.

yeah it's hard to explain, i'll try to create an image later to show it.

**Reply by tsujan** [reply]
> in my example it works fine

I doubt it very much that it would work under all circumstances but feel free to make a PR. I'll be happy to be wrong in this case ;)

> my idea was to have the option but make it clear it "can" cause problems

I understand but, as far as I've seen, ordinary users usually don't pay attention to these things: They see "Unsupported", select it, get surprised that it works ("Why does it work if unsupported?!"), and forget everything. I've really got tired of reports about menus that are opened in wrong screens. The current code prevents such issues.

If we had a single WM (as we should with Wayland), we might support positioning of panels between screens. For now, we have to support all possible WMs and avoid problems.

**Comment by elviosak** [comment]
>I doubt it very much that it would work under all circumstances but feel free to make a PR. I'll be happy to be wrong in this case ;)

[PR sent](https://github.com/lxqt/lxqt-panel/pull/1618), there's some small changes that might need to be done, but i'll leave for u to check it first.

>I understand but, as far as I've seen, ordinary users usually don't pay attention to these things: They see "Unsupported", select it, get surprised that it works ("Why does it work if unsupported?!"), and forget everything. I've really got tired of reports about menus that are opened in wrong screens. The current code prevents such issues.

Understandable, since you're the one dealing with most of those reports..

**Reply by stefonarch** [reply]
> Add another xfce4-whisker-menu style menu, or change current one.

To have the choice between the  "classic" actual menu and an new "improved" would be nice IMO. 
There are some open issues about already:
https://github.com/lxqt/lxqt/issues/413
https://github.com/lxqt/lxqt/issues/255
https://github.com/lxqt/lxqt/issues/100

**Reply by elviosak** [reply]
i'll take a look at those, still need to figure out the existing code.

**Reply by elviosak** [reply]
![menu2](https://user-images.githubusercontent.com/33790211/120726258-0c9e8100-c4ae-11eb-9892-278824282e2b.png)
@stefonarch @tsujan  i worked a bit on making a dual pane menu but i'm still not sure how i'll code it, currently just converting QMenu to dual pane menu, so it's either extra work creating `XdgMenuWidget` and converting it to this new menu, or extra code replicating most of what `XdgMenuWidget` does, i don't quite like either option.

The "Leave" option as a submenu feels weird, but also not sure what to do with it.

I'm planning to just add it to `plugin-mainmenu` with a checkbox on config to choose which menu to use.

As for the suggestions in the linked issues, i think i'll add "Favorites" at first, but only after i replicate the rest.

do you have any suggestions/corrections?

EDIT: the text on Categories buttons are left-aligned, it's just my theme overrides it for some reason.

**Reply by tsujan** [reply]
I think most people that don't like the classic main menu want something like XFCE's start menu (XFCE was mentioned on several occasions).

**Reply by stefonarch** [reply]
Looks good!

> The "Leave" option as a submenu feels weird, but also not sure what to do with it.

How about make the "Leave" button just open lxqt-leave? 
If somebody would like an "hibernate" or "shutdown" button they could add it to the favorites.

Favorites are alphabetical here? Sortable by DND would be nice but not essential I think.

**Reply by elviosak** [reply]
> How about make the "Leave" button just open lxqt-leave?
If somebody would like an "hibernate" or "shutdown" button they could add it to the favorites.

i'll try to make it similar to xfce and make it configurable which buttons to show in a small toolbar probably next to search.

> Favorites are alphabetical here? Sortable by DND would be nice but not essential I think.

DND works with some quirks, the cursor is the copy one, to allow copy/linking to desktop/quicklaunch and NOT allow move fiel, but i might disable (copy/linking) for favorites only, and the other categories will keep the normal behaviour without DND inside it.
i'll end up creating as a separate plugin, i think it will have a lot of different config options and it will work quite differently, so only new one will have at first, i'm slowly writing some pieces, just figured styling for `QListView` to have 2 lines, 

listview styling and DND:
https://user-images.githubusercontent.com/33790211/125178713-b4018880-e1bd-11eb-9066-7974af717c12.mp4
